File: FABRIC STUDIO/lib/credits.ts
import type { GenerationMode, Resolution } from './types';

/**
 * Credit cost matrix. Multiply by num_images.
 *
 *   mode \ res   1k   2k   4k
 *   fast          1    2    3
 *   balanced      2    3    4
 *   quality       3    4    5
 */
const MATRIX: Record<GenerationMode, Record<Resolution, number>> = {
  fast: { '1k': 1, '2k': 2, '4k': 3 },
  balanced: { '1k': 2, '2k': 3, '4k': 4 },
  quality: { '1k': 3, '2k': 4, '4k': 5 },
};

export function creditCost(
  mode: GenerationMode,
  resolution: Resolution,
  numImages = 1,
): number {
  return MATRIX[mode][resolution] * Math.max(1, Math.min(4, numImages));
}

/** Expected wall-clock seconds, used to calibrate the progress readout. */
const LATENCY: Record<GenerationMode, Record<Resolution, number>> = {
  fast: { '1k': 10, '2k': 16, '4k': 24 },
  balanced: { '1k': 16, '2k': 25, '4k': 38 },
  quality: { '1k': 26, '2k': 40, '4k': 55 },
};

export function expectedSeconds(
  mode: GenerationMode,
  resolution: Resolution,
  numImages = 1,
): number {
  const base = LATENCY[mode][resolution];
  // Extra images are batched, not serialised -- they cost roughly 35% each.
  return Math.round(base * (1 + 0.35 * (Math.max(1, numImages) - 1)));
}

/** Preset rungs on the quality control. Advanced can still diverge from these. */
export const QUALITY_PRESETS = [
  {
    id: 'draft',
    label: 'Draft',
    generationMode: 'fast' as GenerationMode,
    resolution: '1k' as Resolution,
    note: '1 MP, about 10s',
  },
  {
    id: 'standard',
    label: 'Standard',
    generationMode: 'balanced' as GenerationMode,
    resolution: '2k' as Resolution,
    note: '4 MP, about 25s',
  },
  {
    id: 'final',
    label: 'Final',
    generationMode: 'quality' as GenerationMode,
    resolution: '4k' as Resolution,
    note: '16 MP, about 55s',
  },
];

export function matchPreset(
  mode: GenerationMode,
  resolution: Resolution,
): string | null {
  const hit = QUALITY_PRESETS.find(
    (p) => p.generationMode === mode && p.resolution === resolution,
  );
  return hit ? hit.id : null;
}

File: studio/credits.py
"""What a generation costs, and roughly how long it takes.

The matrix is FASHN's published `tryon-max` table. Note what is *not* in it:
there is no `fast` mode on `tryon-max` — the endpoint accepts `balanced` and
`quality` only. Sending `fast` does not make a job cheaper, it makes the mode
undefined and lets the API auto-select, which bills as `balanced` while a
naive UI quotes the lower number. So `fast` is not offered.

The `edit` endpoint does accept `fast`, but FASHN does not publish its credit
table, so we quote the `tryon-max` numbers for both rather than invent one.
The true spend comes back from the API in `x-fashn-credits-used` and is what
history records; this table is only ever an estimate shown before the run.

The latency table is ours: it exists to calibrate the progress readout so the
bar moves at a believable rate instead of crawling and then jumping.
"""

MATRIX = {
    "balanced": {"1k": 2, "2k": 3, "4k": 4},
    "quality": {"1k": 3, "2k": 4, "4k": 5},
}

LATENCY = {
    "balanced": {"1k": 20, "2k": 30, "4k": 45},
    "quality": {"1k": 30, "2k": 45, "4k": 70},
}

GENERATION_MODES = ("balanced", "quality")
RESOLUTIONS = ("1k", "2k", "4k")

# The rungs the UI offers. Advanced can still land between them, which is why
# the studio shows "custom" rather than pretending a preset is selected.
QUALITY_PRESETS = [
    {"id": "draft", "label": "Draft", "generationMode": "balanced",
     "resolution": "1k", "note": "1 MP · 2 cr"},
    {"id": "standard", "label": "Standard", "generationMode": "balanced",
     "resolution": "2k", "note": "4 MP · 3 cr"},
    {"id": "final", "label": "Final", "generationMode": "quality",
     "resolution": "4k", "note": "16 MP · 5 cr"},
]


def _clamp_images(num_images):
    try:
        count = int(num_images)
    except (TypeError, ValueError):
        count = 1
    return max(1, min(4, count))


def normalize_mode(mode):
    """`fast` is not a tryon-max mode; it lands on the cheapest real one."""
    return mode if mode in MATRIX else "balanced"


def normalize_resolution(resolution):
    return resolution if resolution in RESOLUTIONS else "1k"


def credit_cost(mode, resolution, num_images=1):
    mode = normalize_mode(mode)
    resolution = normalize_resolution(resolution)
    return MATRIX[mode][resolution] * _clamp_images(num_images)


def expected_seconds(mode, resolution, num_images=1):
    """Extra images are batched rather than serialised — about 35% each."""
    base = LATENCY[normalize_mode(mode)][normalize_resolution(resolution)]
    return int(round(base * (1 + 0.35 * (_clamp_images(num_images) - 1))))


def match_preset(mode, resolution):
    for preset in QUALITY_PRESETS:
        if preset["generationMode"] == mode and preset["resolution"] == resolution:
            return preset["id"]
    return None
